packages/core/src/StateNode.ts

```diff
@@ -954,14 +955,22 @@ class StateNode<
 
     const isInternal = !!selectedTransition.internal;
 
-    const reentryNodes = isInternal
-      ? []
-      : flatten(allNextStateNodes.map((n) => this.nodesFromChild(n)));
+    const reentryNodes: AnyStateNode[] = [];
+
+    if (!isInternal) {
```

Comment:
Note that this algorithm is also broken in other subtle ways. States are not only reentered for external transitions - internal/external only change **what** states are potentially reentered.

For instance, on [this codesandbox](https://codesandbox.io/s/beautiful-benji-njkixn?file=/src/index.js:298-331) this entry action should technically be called twice. Even though the transition is internal, this target state should be reentered.

And I'm not saying that we should fix this right now. I'm just observing the fact. I somewhat don't see how fixing **non-critical** bugs in v4 gets us closer to anything. As mentioned before - I would rather halt the v4 development (just fix pressing issues) and focus on v5 (or on the "satellite" packages, as that development is mostly orthogonal)